hal/src/main/native/athena/HAL.cpp

```diff
@@ -224,14 +219,23 @@ uint64_t HAL_GetFPGATime(int32_t* status) {
     *status = NiFpga_Status_ResourceNotInitialized;
     return 0;
   }
-  std::lock_guard<std::mutex> lock(timeMutex);
-  uint32_t fpgaTime = global->readLocalTime(status);
+  uint64_t upper1 = global->readLocalTimeUpper(status);
+  uint32_t lower = global->readLocalTime(status);
+  uint64_t upper2 = global->readLocalTimeUpper(status);
   if (*status != 0) return 0;
-  // check for rollover
-  if (fpgaTime < prevFPGATime) ++timeEpoch;
-  prevFPGATime = fpgaTime;
-  return static_cast<uint64_t>(timeEpoch) << 32 |
-         static_cast<uint64_t>(fpgaTime);
+  if (upper1 == upper2) {
+    // Did not roll over, either option is viable
+    upper2 = upper2 << 32;
+    upper2 += lower;
+    return upper2;
+  } else {
+    // Rolled over between the lower call, reread lower
+    lower = global->readLocalTime(status);
+    if (*status != 0) return 0;
+    upper2 = upper2 << 32;
+    upper2 += lower;
+    return upper2;
+  }
```

Comment:
We could substantially simplify these branches, and basically just do if (upper1 != upper2) update lower, followed by the rest?

Reply:
Ok fixed again.